src/SOS/Strike/clrma/exception.cpp

```diff
@@ -0,0 +1,551 @@
+            {
+                ArrayHolder<WCHAR> typeName = new (std::nothrow)WCHAR[cbTypeName];
+                if (typeName != nullptr)
+                {
+                    if (SUCCEEDED(hr = m_managedAnalysis->SosDacInterface()->GetMethodTableName(objData.MethodTable, cbTypeName, typeName, nullptr)))
+                    {
+                        m_typeName = typeName.Detach();
+                    }
+                    else
+                    {
+                        TraceError("ClrmaException::Initialize GetMethodTableName(%016x) 2 FAILED %08x\n", objData.MethodTable, hr);
+                    }
+                }
+            }
+            else
+            {
+                TraceError("ClrmaException::Initialize GetMethodTableName(%016x) 1 FAILED %08x\n", objData.MethodTable, hr);
+            }
+            if (m_exceptionData.Message == 0)
+            {
+                // To match the built-in SOS provider that scrapes !pe output.
+                m_message = new (std::nothrow)WCHAR[8];
+                wcscpy(m_message, L"<none>");
+            }
+            else
+            {
+                m_message = m_managedAnalysis->GetStringObject(m_exceptionData.Message);
+            }
+        }
+        m_exceptionDataInitialized = true;
+    }
+    return S_OK;
+}
+
+// IUnknown
+_Use_decl_annotations_
+STDMETHODIMP
+ClrmaException::QueryInterface(
+    REFIID InterfaceId,
+    PVOID* Interface
+    )
+{
+    if (Interface == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+
+    *Interface = nullptr;
+
+    if (IsEqualIID(InterfaceId, IID_IUnknown))
+    {
+        *Interface = (IUnknown*)this;
+        AddRef();
+        return S_OK;
+    }
+    else if (IsEqualIID(InterfaceId, __uuidof(ICLRMAClrException)))
+    {
+        *Interface = (ICLRMAClrException*)this;
+        AddRef();
+        return S_OK;
+    }
+
+    return E_NOINTERFACE;
+}
+
+STDMETHODIMP_(ULONG)
+ClrmaException::AddRef()
+{
+    return (ULONG)InterlockedIncrement(&m_lRefs);
+}
+
+STDMETHODIMP_(ULONG)
+ClrmaException::Release()
+{
+    LONG lRefs = InterlockedDecrement(&m_lRefs);
+    if (lRefs == 0)
+    {
+        delete this;
+    }
+    return lRefs;
+}
+
+//
+// ICLRMAClrException
+//
+
+_Use_decl_annotations_
+STDMETHODIMP
+ClrmaException::get_DebuggerCommand(
+    BSTR* pValue
+    )
+{
+    if (pValue == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+    *pValue = nullptr;
+    return E_NOTIMPL;
+}
+
+_Use_decl_annotations_
+STDMETHODIMP
+ClrmaException::get_Address(
+    ULONG64* pValue
+    )
+{
+    if (pValue == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+    *pValue = m_address;
+    return S_OK;
+}
+
+_Use_decl_annotations_
+STDMETHODIMP
+ClrmaException::get_HResult(
+    HRESULT* pValue
+    )
+{
+    if (pValue == nullptr)
+    {
+        return E_INVALIDARG;
+    }
+
+    *pValue = 0;
+
+    HRESULT hr;
+    if (FAILED(hr = Initialize()))
```

Comment:
nit: consider IfFailRet macro

Reply:
I really dislike that macro. It assumes (at the version in Strike.cpp) that the HRESULT is called Status. I like to see the explicit return.  And it isn't available in the h file this code includes. I'm really trying keep this new code clean of these kind of macros except FAILED/SUCCEEDED.